### PR TITLE
[language] beginning of a Debug module

### PR DESCRIPTION
**Pull request by vgao1996**
## Summary
This adds a module `Debug` with a native function `print`. `print` takes a reference to a value of any type and prints out the value in stdout. 

Highlights:
- Implementation of the native function `print` is behind a feature flag
  - No-op if feature flag is not present
- IR tests now use freshly generated genesis
- Move-lang functional tests now run faster

## Test Plan
Run `cargo test -p move-lang debug/simple.move` manually and you should see
```
[debug] 42u64
[debug] [100u64, 200u64, 300u64]
[debug] M::Foo {false}
[debug] M::Bar {404u128, M::Foo {false}, true}
[debug] M::Box {M::Foo {false}}
```

## Progress Tracker
- [x] Basic Functionalities
- [x] Documentation
- [x] Conditional Compilation
- [x] Debug/Prod Genesis & Tests
- [x] Code Cleanup

### CODE REVIEW

**Comment by dariorussi** [comment]
let's talk about this when we have a chance because there are few things I am not sure I understand:
- why not having a standard native function there? I mean why having the interpreter specializing the print function?
- we want a generic ability to print values. We have that API already lying around a bit (used to be pretty_print or something) and it seems we want that.
Also that is what a `print_stack_trace` and the core dump could use.
Thoughts?

**Comment by tnowacki** [comment]
Ditto to all of @dariorussi's comments
- I think these should be normal native functions
- +1 to print_stack_trace

**Comment by vgao1996** [comment]
@dariorussi @tnowacki Yes you've made a good point and it's also something I'd like to discuss specifically! To answer both of your questions: I thought it would be more user friendly if the information being printed out could contain some type info (struct name, vec vs struct etc.), so I tried to pass in the nominal types and an environment in which you can lookup the layout.

My bad to not have made this clear beforehand, but think of my current implementation as an experiment and use it to see the trade-offs: in order to include the type info we'll have to complicate our code and implement the print function in the interpreter. 

Personally I don't think this is buying us much at this time, considering the VM will be refactored drastically, but what do you guys think?

**Comment by tnowacki** [comment]
> To answer both of your questions: I thought it would be more user friendly if the information being printed out could contain some type info (struct name, vec vs struct etc.), so I tried to pass in the nominal types and an environment in which you can lookup the layout.

There is a `Vec<TypeTag>` passed into the native function invocation. Can't you use that to provide that information?

> Personally I don't think this is buying us much at this time, considering the VM will be refactored drastically, but what do you guys think?

Could you rephrase? I'm not sure what is buying what :p

**Comment by vgao1996** [comment]
> There is a Vec<TypeTag> passed into the native function invocation. Can't you use that to provide that information?
Ah I'm stupid. With both the `Vec<TypeTag>` and `Vec<Type>` (aka layout) being passed in I will actually be able to tell the names of nested structs. Thanks for catching this!

> Could you rephrase? I'm not sure what is buying what :p
Complex + ugly code vs being able to differentiate values of different types, but now we can see it's a non-issue.

**Comment by vgao1996** [comment]
Update: rebase and use the new `Type` so that `print` is implemented as a normal native function.

File: language/move-vm/runtime/src/special_names.rs [hunk 1]
```diff
@@ -11,3 +11,5 @@ pub(crate) static EMIT_EVENT_NAME: Lazy<Identifier> =
 
 pub(crate) static SAVE_ACCOUNT_NAME: Lazy<Identifier> =
     Lazy::new(|| Identifier::new("save_account").unwrap());
+
+pub(crate) static PRINT_NAME: Lazy<Identifier> = Lazy::new(|| Identifier::new("print").unwrap());
```
Comment (dariorussi) [inline language/move-vm/runtime/src/special_names.rs:15]: is this needed at all? I ask because I don't see any usage and I would expect no usage to be there

File: language/move-vm/types/src/values/values_impl.rs [hunk 2]
```diff
@@ -1884,6 +1884,188 @@ impl Display for Locals {
     }
 }
 
+pub mod debug {
```
Comment (dariorussi) [inline language/move-vm/types/src/values/values_impl.rs:1887]: this is all good but I would really think about how to make all of this just the printing/display utilities for move values.
In other words the debug module should just be a call into a "display" function.
I think we want that and the stack/environment (core dump) printing as a core part of our display/debug story and it would be nice to see all of that work together.
Those are considerations and no particular action to take here

File: language/move-vm/runtime/src/special_names.rs (hunk 1 above)
Comment (vgao1996) [reply language/move-vm/runtime/src/special_names.rs:15]: No it's not. Forgot to delete it.

File: language/move-vm/types/src/values/values_impl.rs (hunk 2 above)
Comment (vgao1996) [reply language/move-vm/types/src/values/values_impl.rs:1887]: So here we have two kinds of print/Display functions: debug::print with Type (the one you are seeing right now) and Display without Type. My understanding is that debug::print is facing Move developers and should be optimized for user experience. The rust Debug/Display dumps are more for ourselves to inspect a VM crash or whatever. Now what's unclear to me is that once we have debug::print, do we still need Debug/Display.

**Comment by vgao1996** [comment]
Update: 
- Done with conditional compilation
- Added a `//! debug` config option to functional tests
- Switched the IR testsuite to use freshly generated genesis

The code is a little bit messy right now as there's an issue with the way Move functional tests handle dependencies and I'll need to clean this up a bit. 

There's also a regression on the running time of IR tests. This is because it now needs to actually execute the genesis transaction for every test, but I know how to fix it.

File: language/move-vm/types/src/values/values_impl.rs [hunk 3]
```diff
@@ -1884,6 +1884,188 @@ impl Display for Locals {
...
+    use super::*;
```
Comment (sblackshear) [inline language/move-vm/types/src/values/values_impl.rs:1888]: This module feels like it could/should live in its own file unless the visibility rules of `values_impl` prevent it.

Comment (vgao1996) [reply language/move-vm/types/src/values/values_impl.rs:1888]: The current visibility rules require this to be a sub-module of values_impl, but yeah it can live in its own file. I'll take a look at a later time and see what will be the best module & directory structure for move_vm/types.

**Comment by vgao1996** [comment]
@sblackshear the reason why I'm championing for conditional compilation in this case is that the debug module is actually very very powerful: 
- It has access to the standard output of the vm, which as we all know, if not handled properly, can enable all sorts of attacks.
- The debug functions do NOT consume gas, in order to minimize its visible side effects.
- It inspects the internals of the values and types. This process is already NOT free of side-effects due to our having `RefCell`s. It will be even more dangerous if we introduce some lazy structures in the future.
- Although not implemented yet, we will have another function printing out the stack traces. This is also a very dangerous operation as it traverses the whole call stack. Again, don't consider this a read-only operation.

Having these code in the TCB binary at all will make it considerably harder to audit or at least require more work, so I'd rather kill this off completely in the production binary.

**Comment by vgao1996** [comment]
Regarding the on-chain and local standard libraries, my understanding is that if a native function is  not available in the VM, then it must not be in the genesis, or otherwise we won't be able to replay the genesis transaction. (There's some complexities here, but the stdlib must verify and link.) Is this actually a problem @sblackshear & @runtian-zhou ?

One alternative approach is to keep a declaration of `Debug::print` in the VM, but make it a runtime error if one ever tries to execute it. (It must still verify and link.) This will still require conditional compilation or passing down a flag, which I don't think we'll do.

I totally agree that `debug_module: bool` won't scale well, but given how the current code is structured if we want to have a more comprehensive story for genesis management I'd say we'll rewrite that part significantly anyways.

File: language/move-vm/types/src/values/values_impl.rs (hunk 2 above)
Comment (dariorussi) [reply language/move-vm/types/src/values/values_impl.rs:1903]: I'd say we want a unified story for all kind of printing and let client decide what they want to use. And by client I mean internally us. That is, we call the most sensible API for whatever it is that we are doing.
Again though I don't think we should do it in this diff but it seems a good thing to have to me.
Thoughts?

**Comment by sblackshear** [comment]
@vgao1996: you make some compelling arguments for why `Debug` is scarier than one might expect. But these only apply if it can be invoked in production. We will not call it from any stdlib module or stdlib script, nor will we allow custom scripts that do so. I think much more caution would be warranted in a system where an attacker can write arbitrary code (for all of the reasons you mention). But I also think having `Debug` exposed would be a minor worry in such a system compared to (e.g.) hardening the bytecode verifier.

> Having such code in the TCB binary at all will make it considerably harder to audit or at least require more work, so I'd rather kill it off completely in the production binary.

Auditing is done at source level at the granularity of of crates, not individual functions or structs. So I don't think the conditional compilation will make much of a difference here.

> One alternative approach is to keep a declaration of Debug::print in the VM, but make it a runtime error if one ever tries to execute it. (It must still verify and link.) This will still require conditional compilation or passing down a flag, which I don't think we'll do.

I like this approach (conditional compilation inside the native impl of Debug that's a no-op or abort) much better than having `debug_module` configs for both the Move standard library and VM that can get out of sync + make the code nastier (on the stdlib side). But overall I think having no configs for now and considering closer to shipping (as @dariorussi suggests) will cause us the least pain with very little risk given the absence of arbitrary code.

**Comment by liangping** [comment]
Hey, I am later on this topic, I really think this a useful tool for Move developers. and I did a similar implementation called "Console.log". 

One thing to be added:
```
[debug] 42u64   // better to print like "x=42u64"
[debug] [100u64, 200u64, 300u64] // constant=[100u64, 200u64,300u64]
[debug] M::Foo {false}
[debug] M::Bar {404u128, M::Foo {false}, true}
[debug] M::Box {M::Foo {false}}
```

**Comment by vgao1996** [comment]
@sblackshear @dariorussi Yeah agreed it's a bit too early to think about conditional compilation and I'll get that part removed. Thanks for pointing this out!

I'll remove the debug_module flag for now.

**Comment by vgao1996** [comment]
@liangping Thanks for the suggestion! Right now I don't think it is possible to print out variable names, as they are not available in the current file format. (It will be possible if the VM takes auxiliary debug info, but that will be a more substantial change.)

In addition, we will add another function to allow you to print out stack traces, which will include assignments to local variables.
